src/python/pants/backend/python/lint/helpers.py

```diff
@@ -0,0 +1,112 @@
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+
+@dataclass(frozen=True)
+class ToolSetup:  # @TODO: probably needs a better name
+    process: Process
+    original_digest: Digest
+
+
+# ============================
+async def setup_tool(
+    setup_request,
+    python_tool: PythonToolBase,
+    interpreter_constraints: InterpreterConstraints,
+    get_argv,
+    get_config=None,
+) -> ToolSetup:
+    pex_get = Get(
+        VenvPex,
+        PexRequest(
+            output_filename=f"{python_tool.options_scope}.pex",
+            internal_only=True,
+            requirements=python_tool.pex_requirements(),
+            interpreter_constraints=interpreter_constraints,
+            main=python_tool.main,
+        ),
+    )
+
+    source_files_get = Get(
+        SourceFiles,
+        SourceFilesRequest(field_set.source for field_set in setup_request.request.field_sets),
+    )
+
+    source_files, black_pex = await MultiGet(source_files_get, pex_get)
+    source_files_snapshot = (
+        source_files.snapshot
+        if setup_request.request.prior_formatter_result is None
+        else setup_request.request.prior_formatter_result
+    )
+
+    input_digest = source_files_snapshot.digest
+    if get_config:
+        config_files = await get_config(source_files_snapshot)
+        input_digest = await Get(
+            Digest, MergeDigests((source_files_snapshot.digest, config_files.snapshot.digest))
+        )
+
+    process = await Get(
+        Process,
+        VenvPexProcess(
+            black_pex,
+            argv=get_argv(source_files, python_tool, check_only=setup_request.check_only),
+            input_digest=input_digest,
+            output_files=source_files_snapshot.files,
+            description=f"Run {python_tool.options_scope.title()} on {pluralize(len(setup_request.request.field_sets), 'file')}.",
+            level=LogLevel.DEBUG,
+        ),
+    )
+    return ToolSetup(process, original_digest=source_files_snapshot.digest)
+
+
+# ====
+
+
+async def run_formatter_from_venv(
+    subsystem: Subsystem,
+    # NB: We don't use `setup_tool` directly so the user can write a rule to get memoization
+    get_setup,
+) -> FmtResult:
+    name = subsystem.options_scope.title()
+    if getattr(subsystem, "skip", False):
+        return FmtResult.skip(formatter_name=name)
+    setup = await get_setup()
+    result = await Get(FallibleProcessResult, Process, setup.process)
+    return FmtResult.from_process_result(
+        result,
+        original_digest=setup.original_digest,
+        formatter_name=name,
+    )
+
+
+# ====
+
+
+async def run_linter_from_venv(
+    subsystem: Subsystem,
+    # NB: We don't use `setup_tool` directly so the user can write a rule to get memoization
+    get_setup,
+) -> LintResults:
+    name = subsystem.options_scope.title()
+    if subsystem.skip:
+        return FmtResult.skip(formatter_name=name)
+    setup = await get_setup()
```

Comment:
Why this level of indirection for `get_setup`?

Rather than `run_linter_from_venv(..., get_setup=Get(ToolSetup, SetupRequest, ..))` ?

Reply:
So why is it a callable? I just didn't think of that when I wrote it :smile: 

.... I mean, uh, I just wanted to demonstrate the parser parses lambdas :eyes: